engine/runtime/src/main/java/org/enso/interpreter/runtime/ReferencesManager.java

```diff
@@ -0,0 +1,58 @@
+package org.enso.interpreter.runtime;
+
+import java.lang.ref.Reference;
+import java.lang.ref.ReferenceQueue;
+import java.lang.ref.SoftReference;
+import java.lang.ref.WeakReference;
+import java.util.Collection;
+import java.util.concurrent.ConcurrentLinkedQueue;
+
+/** Tracks soft and weak references and allow their cleanup. */
+public final class ReferencesManager {
+  private final EnsoContext ctx;
+  private final Collection<Reference> refs = new ConcurrentLinkedQueue<>();
+  private final ReferenceQueue<Object> queue = new ReferenceQueue<>();
+
+  ReferencesManager(EnsoContext ctx) {
+    this.ctx = ctx;
+  }
+
+  /**
+   * Creates new reference to provided object and registers it in the manager.
+   *
+   * @param <T> class of the object to reference
+   * @param obj the object to reference
+   * @param type ({@code 1} use {@link SoftReference} or {@code 2} to use {@link WeakReference}
+   * @return newly created reference to the provided object
+   */
+  public <T> Reference<T> create(T obj, int type) {
```

Comment:
One of the future directions is to _enhance the caching_ and make it more _"intelligent"_. To do so, we could:
```java
public <T> Reference<T> create(T obj, long sizeInBytes, long timeInMillisToCompute)
```
e.g. let every reference be associated with its **size** and **time to compute**. Based on that information the `ReferencesManager` could select appropriate _caching strategy_.